Sources/SimilaritySearchKit/Core/Embeddings/Models/NativeContextualEmbeddings.swift

```diff
@@ -0,0 +1,89 @@
+//
+//  NativeContextualEmbeddings.swift
+//  
+//
+//  Created by Zach Nagengast on 10/11/23.
+//
+
+import Foundation
+import NaturalLanguage
+import CoreML
+
+#if canImport(NaturalLanguage.NLContextualEmbedding)
+@available(macOS 14.0, iOS 17.0, *)
+public class NativeContextualEmbeddings: EmbeddingsProtocol {
+    public let model: ModelActor
+    public let tokenizer: any TokenizerProtocol
+
+    // Initialize with a language
+    public init(language: NLLanguage = .english) {
+        self.tokenizer = NativeTokenizer()
+        guard let nativeModel = NLContextualEmbedding(language: language) else {
+            fatalError("Failed to load the Core ML model.")
+        }
+        Self.initializeModel(nativeModel)
+        self.model = ModelActor(model: nativeModel)
+    }
+
+    // Initialize with a script
+    public init(script: NLScript) {
+        self.tokenizer = NativeTokenizer()
+        guard let nativeModel = NLContextualEmbedding(script: script) else {
+            fatalError("Failed to load the Core ML model.")
+        }
+        Self.initializeModel(nativeModel)
+        self.model = ModelActor(model: nativeModel)
+    }
+
+    // Common model initialization logic
+    private static func initializeModel(_ nativeModel: NLContextualEmbedding) {
+        if !nativeModel.hasAvailableAssets {
+            nativeModel.requestAssets { _, _ in }
+        }
+        try! nativeModel.load()
+    }
+
+    // MARK: - Dense Embeddings
+
+    public actor ModelActor {
+        private let model: NLContextualEmbedding
+
+        init(model: NLContextualEmbedding) {
+            self.model = model
+        }
+
+        func vector(for sentence: String) -> [Float]? {
+            // Obtain embedding result for the given sentence
+            // Shape is [1, embedding.sequenceLength, model.dimension]
+            let embedding = try! model.embeddingResult(for: sentence, language: nil)
+
+            // Initialize an array to store the total embedding values and set the count
+            var meanPooledEmbeddings: [Float] = Array(repeating: 0.0, count: model.dimension)
+            let sequenceLength = embedding.sequenceLength
+
+            // Mean pooling: Loop through each token vector in the embedding and sum the values
+            embedding.enumerateTokenVectors(in: sentence.startIndex ..< sentence.endIndex) { (embedding, _) -> Bool in
+                for tokenEmbeddingIndex in 0 ..< embedding.count {
+                    meanPooledEmbeddings[tokenEmbeddingIndex] += Float(embedding[tokenEmbeddingIndex])
+                }
+                return true
+            }
+
+            // Mean pooling: Get the average embedding from totals
```

Comment:
I'm new to all this - is this the best practice or are there alternatives? appreciate any pointers on what to read into. thanks

Reply:
You can use max pooling, or some other method - see sentence transformers https://www.sbert.net/examples/applications/computing-embeddings/README.html?highlight=pooling#sentence-embeddings-with-transformers. I've been told mean pooling is the best option for sentences.

Reply:
thanks. I've read up a little more now.

so to understand - you expect the lib user to take user input (for example), split it into sentences (via user-provided tokenization) and then each sentence gets a vector based on the average of each token within the sentence?

is it possible to do more granular mean pooling, so that a sentence can be matched via a few possible more specific vectors? or does that not matter because the contextualized vectors already embed the full sentence (and beyond the sentence) into each individual vector to some degree?

Reply:
The lib was designed with extensibility in mind, but there are several helpers to assist with what you mentioned, but its up to you to pick the best method for your dataset because some techniques will work better than others depending on what your putting through.